src/fastoad/openmdao/problem.py

```diff
@@ -176,68 +231,28 @@ def _read_inputs_without_setup_done(self):
-            # in input file.
-            ivc_vars = tmp_prob.model.get_io_metadata(
-                "output",
-                tags=["indep_var", "openmdao:indep_var"],
-                excludes=f"{SHAPER_SYSTEM_NAME}.*",
-            )
-        for meta in ivc_vars.values():
-            try:
-                del input_variables[meta["prom_name"]]
-            except ValueError:
-                pass
+
+        input_variables = VariableList(
+            [
+                variable
+                for variable in input_variables
+                if variable.name not in self._analysis.ivc_var_names
+            ]
+        )
+
         if input_variables:
+            self._analysis.dynamic_input_vars = VariableList(
+                [
+                    variable
+                    for variable in self._analysis.dynamic_input_vars
+                    if variable.name not in input_variables.names()
+                ]
+            )
             self._insert_input_ivc(input_variables.to_ivc())
 
     def _insert_input_ivc(self, ivc: om.IndepVarComp, subsystem_name=INPUT_SYSTEM_NAME):
-        with problem_without_mpi(self) as tmp_prob:
-            tmp_prob.setup()
-
-            # We get order from copied problem, but we have to ignore the "shaper"
-            # and the auto IVCs.
-            previous_order = [
-                system.name
-                for system in tmp_prob.model.system_iter(recurse=False)
-                if system.name != "_auto_ivc" and system.name != SHAPER_SYSTEM_NAME
-            ]
-
         self.model.add_subsystem(subsystem_name, ivc, promotes=["*"])
-        self.model.set_order([subsystem_name] + previous_order)
-
-    @classmethod
-    def _get_undetermined_dynamic_vars_metadata(cls, problem):
-        """
-        Provides dict (name, metadata) for dynamically shaped inputs that are not
-        fed by an existing output (assuming overall variable promotion).
-
-        Assumes problem.setup() has been run, at least partially.
-
-        :param problem:
-        """
-        # First all outputs are identified. If a dynamically shaped input is fed by a matching
-        # output, its shaped will be determined.
-        output_var_names = []
-        for system in problem.model.system_iter(recurse=False):
-            io_metadata = system.get_io_metadata("output")
-            output_var_names += [meta["prom_name"] for meta in io_metadata.values()]
-
-        dynamic_vars = {}
-        for system in problem.model.system_iter(recurse=False):
-            io_metadata = system.get_io_metadata("input")
-            dynamic_vars.update(
-                {
-                    meta["prom_name"]: meta
-                    for name, meta in io_metadata.items()
-                    if meta["shape_by_conn"] and meta["prom_name"] not in output_var_names
-                }
-            )
-        return dynamic_vars
+        self.model.set_order([subsystem_name] + self._analysis.subsystem_order)
```

Comment:
Again not linked with this PR, but do we need to reorder the subsystems ? Isn't naturaly done when we add the models  ?

Reply:
A sure thing is that the component are added in the order of the `add_subsystem` calls.

The fact is that removing this line makes 2 tests fail (these tests do optimization on the Sellar problem).
Other tests (unit and integration) still work fine. 
For the record, with the last OpenMDAO version installed, using `self.model.options["auto_order"] = True` makes the 2 tests work again.

Actually, the 2 tests fail (give incorrect results) because there is no solver at the `Problem.model` level. Indeed, with an IVC in last position, a loop is needed to have the IVC output correctly fill the needed inputs.